utils/vlog/vlog_to_pbtype.py

```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/python3
+import yosys_exec
+import lxml.etree as ET
+import argparse, re
+import os, tempfile
+from yosys_json import YosysJson
+
+"""
+Convert a Verilog simulation model to a VPR `pb_type.xml`
+
+The following Verilog attributes are allowed on instances:
+    - (* PB_PATH="../path/to/pb_type.xml" *) : import the pb_type for a given instance
+        from an external XML file instead of including it inline.
+    
+    - (* GENERATE_PB *) : generate (and overwrite if applicable) the above file, as well as including it
+        
+The following are allowed on both instances (except where PB_PATH is set but GENERATE_PB isn't), and
+the top level module.
+    
+    - (* BLIF_MODEL=".latch" *) : specify the corresponding blif_model for a given instance
+    
+    - (* CLASS="flipflop" *) : specify the corresponding class for a given instance
+
+    - (* PB_NAME="name" *) : override the name of the pb_type (default: cell type)
+
+The following are allowed on nets within modules:
+
+    - (* SETUP="clk 10e-12" *) : specify setup time for a given clock
```

Comment:
I'm not sure, but I think Verilog has a native way of specifying setup and hold times? 
http://verilog.renerta.com/mobile/source/vrg00052.htm
> Timing Check Tasks are for verification of timing properties of designs and for reporting timing violations.

http://athena.ecs.csus.edu/~changw/class_docs/VerilogManual/me95rh.html
>This section, the final part of the delay modeling chapter, discusses some of the various system tasks that exist for the purposes of timing checks. Verilog contains many timing-check system tasks, but only the three most common tasks are discussed here: $setup, $hold and $width. Timing checks are used to verify that timing constraints are upheld, and are especially important in the simulation of high-speed sequential circuits such as microprocessors. All timing checks must be contained within specify blocks as shown in the example below.

Also think delays are part of verilog too? http://verilog.renerta.com/mobile/source/vrg00011.htm ?
> Delays specify a time in which assigned values propagate through nets or from inputs to outputs of gates.

If these attributes are "just temporary" until we add proper support for these values, then can you add a "FIXME: Use proper verilog for these" or something?

Reply:
It does, but Yosys doesn't output them in the JSON output at the moment. I will discuss this with Clifford, maybe I will look into adding this to Yosys